AccessibilityInsightsForAndroidService/app/src/main/java/com/microsoft/accessibilityinsightsforandroidservice/OnScreenshotAvailable.java

```diff
@@ -31,26 +34,74 @@ public synchronized void onImageAvailable(ImageReader imageReader) {
+      Logger.logError(TAG, "ImageFormatException: " + e.toString());
+      image.close();
+    }
   }
 
-  private Bitmap getBitmapFromImage(Image image) {
-    Image.Plane[] imagePlanes = image.getPlanes();
-    int bitmapWidth = getBitmapWidth(image, imagePlanes);
-    Bitmap screenshotBitmap =
-        bitmapProvider.createBitmap(bitmapWidth, metrics.heightPixels, Bitmap.Config.ARGB_8888);
-    ByteBuffer buffer = imagePlanes[0].getBuffer();
-    screenshotBitmap.copyPixelsFromBuffer(buffer);
-    return screenshotBitmap;
+  private Bitmap getBitmapFromImage(Image image) throws ImageFormatException {
+    int width = image.getWidth();
+    int height = image.getHeight();
+    if (width != metrics.widthPixels || height != metrics.heightPixels) {
+      Logger.logError(
+          TAG,
+          "Received image of dimensions "
+              + width
+              + "x"
+              + height
+              + ", mismatches device DisplayMetrics "
+              + metrics.widthPixels
+              + "x"
+              + metrics.heightPixels);
+    }
+
+    Bitmap bitmap = bitmapProvider.createBitmap(width, height, IMAGE_BITMAP_FORMAT);
+    copyPixelsFromImagePlane(bitmap, image.getPlanes()[0], width, height);
+
+    return bitmap;
   }
 
-  private int getBitmapWidth(Image image, Image.Plane[] imagePlanes) {
-    int pixelStride = imagePlanes[0].getPixelStride();
-    int rowStride = imagePlanes[0].getRowStride();
-    int rowPadding = rowStride - pixelStride * metrics.widthPixels;
-    return image.getWidth() + rowPadding / pixelStride;
+  private void copyPixelsFromImagePlane(
+      Bitmap destination, Image.Plane source, int width, int height) throws ImageFormatException {
+    // Note: rowStride is usually pixelStride * width, but can sometimes be larger (with padding)
+    int pixelStride = source.getPixelStride(); // bytes per pixel
+    int rowStride = source.getRowStride(); // bytes per row
+
+    if (pixelStride != IMAGE_PIXEL_STRIDE) {
+      throw new ImageFormatException(
+          "Invalid source Image: pixelStride=" + pixelStride + ", expected " + IMAGE_PIXEL_STRIDE);
+    }
+    if (rowStride < width * pixelStride) {
+      throw new ImageFormatException(
+          "Invalid source Image: rowStride "
+              + rowStride
+              + " is too small for width "
+              + width
+              + " at pixelStride "
+              + pixelStride);
+    }
+
+    ByteBuffer sourceBuffer = source.getBuffer();
+
+    if (rowStride == width * pixelStride) {
+      destination.copyPixelsFromBuffer(sourceBuffer);
+      return;
+    }
+
+    int unpaddedRowStride = width * pixelStride;
+    ByteBuffer pixelDataWithRowPaddingRemoved = ByteBuffer.allocate(unpaddedRowStride * height);
+    for (int row = 0; row < height; ++row) {
+      sourceBuffer.position(row * rowStride);
+      sourceBuffer.get(
+          pixelDataWithRowPaddingRemoved.array(), row * unpaddedRowStride, unpaddedRowStride);
```

Comment:
Potentially dumb question: Does the target (`pixelDataWithRowPaddingRemoved.array()`) automatically get updated by between rows? It seems like this would copy each row into the same chunk of memory. The docs for `ByteBuffer.get(byte[] dst, int offset, int length)` have no reference to how the `dst` buffer gets updated--it would work if it called the `put` method, but the docs are mute on if that's the case...

Reply:
Not dumb at all, I found it to be really confusing exactly when/how the positions of the `ByteBuffer`s get updated.

This code is (intentionally) not updating `pixelDataWithRowPaddingRemoved`'s position; the `offset` parameter of `get` is the offset into the `dst` array that it's writing to, not the offset into the source buffer that it's reading from (the offset into the source buffer is handled by the `position` call on the previous line)

Reply:
Refactored a bit to try to make this more clear from the variable names/comments!